### ISSUE TITLE
[2019] RoBERTa: A Robustly Optimized BERT Pretraining Approach

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#31]
## 論文リンク
https://arxiv.org/abs/1907.11692

## 公開日（yyyy/mm/dd）
2019/07/26

## 概要
あとで書く。

**Comment by yoheikikuta** [comment]
BERT 以降 Transformer-XL や XLNet など色々な改良がなされてきたが、そもそも BERT の能力を十分に発揮すれば BERT が一番、という話を聞いて実際に論文を読んでみようと思い立った。

**Comment by yoheikikuta** [comment]
全体的な印象としては論文というよりはテクニカルレポート。
BERT の性能を理解するために色々な要素に注目して追試実験なりをしてみたら、実は original の BERT よりも高い性能を発揮することがわかったので論文としてまとめた、という感じだろうか。

主張している論文の貢献は以下の三つ。

- BERT のモデルデザインに関して downstream task の performance を高めるために効く要素を同定
- （これまではやられていない）private でないデータセット CC-NEWS を使ってデータ量に対する振る舞いを検証
- 提案する学習によって BERT 以降に提案されたモデルと同等以上の性能を発揮することを示した

**Comment by yoheikikuta** [comment]
Background に関しては BERT の復習という感じ。
これは BERT を真面目に読んだことがあればわかっている話だが、後々重要になる点としては以下。

- Masked Language Model は学習前にデータを作ってそれを使い回す。ただしデータは複製するので同じ文章でも違う mask がされたデータが生成されている
- Next Sentence Prediction (NSP) は文章間の関係を学習するための pre-training で、特に Natural Language Inference のような downstream task の performance 向上に寄与
- 使っているデータは BOOKCORPUS と WIKIPEDIA で uncompressed text としては 16[GB] というサイズ感

**Comment by yoheikikuta** [comment]
まず original の BERT の追試実験やったら Adam の epsilon term に sensitive とか $` β_2 `$ を 0.999 -> 0.98 に変えた方が安定とかなかなかに細かい。

少し大きめの違いとしては max length を 512 で固定してずっと学習している点。
original では 128 とかで最初の方に学習してから 512 に増やしているが、そういうことはしていないとのこと。
これに関して performance にどう影響があったかについては言及がないが、自分で触ってみた感じこの影響はあまりなさそうな印象だし、特別重要ではないということなんだろう。

ちなみにこの論文では BERT のアーキテクチャに関してはいじらない。
学習パラメタに関してもバッチサイズとステップ数以外は変更しない（上で発見した β_2 = 0.98 は使ってる）。

データは BOOKCORPUS + WIKIPEDIA, CC-NEWS, OPENWEB TEXT, STORIES を集めてきて合計 160[GB] とのこと。自然言語処理でこのサイズはなかなかにデカイね。

**Comment by yoheikikuta** [comment]
評価で使ったタスクは GLUE, SQuAD, RACE の三つ。
前者二つは original の BERT の論文と同じだし、どうやって finetuning して解くかも同様なのでここでは割愛する。

[RACE](https://arxiv.org/abs/1704.04683) に関しては知らなかったので軽く調べてみた。

```
Collected from the English exams for middle and high school Chinese students in the age
range between 12 to 18, RACE consists of near 28,000 passages and near 100,000 
questions generated by human experts (English instructors), 
and covers a variety of topics which are carefully designed for evaluating 
the students’ ability in understanding and reasoning.
```

具体的なサンプルが以下。確かになかなか長めの passage となっていることがみて取れる。

<p align="center">
  <img src="https://i.imgur.com/igwwmxy.png">
</p>

正しい選択肢を選ぶ、という問題なので解き方としては全選択肢それぞれの問題文と concat と passage のペア予測をして最も確率の高いものにする、とかになるのかな。異なるペア予測の出力は比較できないのでうーんという感じはするが。

**Comment by yoheikikuta** [comment]
準備は整ったので各種実験とその結果を見ていく。

まずは mask に関して。
この論文では input feed のたびに mask を変える dynamic mask を提案している。この場合毎回異なる mask が得られるが都度変換をしないといけない。
結果は以下で、reference が original 論文の数値。ちょっとは良くなっているかもしれないがそこまで効いてるというわけでもない。個人的にはこれなら static の方が速度の面も含めて良いのではないかと思う。

<p align="center">
  <img src="https://i.imgur.com/PBsiW20.png">
</p>

次に NSP に関して。
NSP は入れないと性能が下がるとか入れなくても変わらんとか色々主張があるので、この論文ではいくつかのパターンを検証してその違いを評価。
まず入力は複数文章からなる segment を二つ入力とするパターンと、二つの文章を入力とする（つまり短い）パターンを NSP 有りで検証。
さらに NSP 無しの場合は、document をまたぐことを気にせずに 512 token ギチギチに詰める full-sentences と document またぎが発生する場合はそこで打ち切る doc-sentences の場合を検証。

結果としては、NSP なしでも性能は落ちないもしくは向上する、というもの。
まあこの辺はタスク依存性とか色々な条件がありそうだし、絶対に NSP はいらん！というほどの主張には成り難い気はする。
一番良いのは doc-sentences だが、token の数が変わったりしてメモリを使い切るのが大変なので、token 数を fix できる full-sentences を使っていく、と言っている。

<p align="center">
  <img src="https://i.imgur.com/KUYeuSa.png">
</p>

あとはバッチサイズを変えた結果。learning rate も合わせて tune すれば結構デカいバッチサイズでも良いとのこと。結構 perplexity の値が変わる。task performance はそこまで変わらんが。

<p align="center">
  <img src="https://i.imgur.com/zohHOXE.png">
</p>

**Comment by yoheikikuta** [comment]
そして結構複雑な text encoding の話。

まず、Byte-Pair Encoding (BPE) というサブワード単位で分割してその単位で embedding を学習するというものが結構使われているが、何を単位にしているかは注意が必要になる。

例えば BERT の WordPiece embedding は、文字を単位にして良く出てくる 't' 'h' 'e' の並びをまとめるとかそういう処理をしている。文字コードの文脈で言えば、ユニコードのコードポイントに割り当てられる文字を単位にしているということになる。なので漢字の '魁' なども一単位として扱われることになる。

もし全ての文字を対象にしようと思ったら、BPE ではまずその最小単位のものを全て登録するので、何十万とかいう最小単位を登録し、その上で良く出現する組み合わせをさらに登録していくことになるため、辞書が膨大になってしまう。

それゆえに、英語だけとか日本語だけとかで使えるモデルになってしまう。

原理的には、例えば UTF-8  encoding を使うと定めてしまえば、全ての文字は [0, 255] のバイトの羅列として捉えることができるので、"真の" BPE としてこれを採用することもできる。
`Language Models are Unsupervised Multitask Learners` の論文ではなんでもかんでも merge しようとすると望ましくないので unicode character category の同じものだけを merge 対象にすると言ってたりもする。

この論文では、その中間のようなことをしていて、雑に言えば ASCII コード表に含まれる文字は文字単位で取り扱ってそれ以外は byte 単位で扱うということをしている。英語はちゃんと英語として扱うよということですかね。

以上により、一応日本語も同じ枠組みで encoding することができる。
例としては以下。

```
roberta.encode('あ'), roberta.encode('日'), roberta.encode('魁')

(tensor([    0, 49260,     2]),
 tensor([    0, 47954,  8210,     2]),
 tensor([    0, 41907, 12410, 10172,     2]))
```

0, 2 は start と end なのでどの入力に対しても付与される。
UTF-8 では日本語はだいたい 3 バイトなので、あまり出てこないパターンだと 3 token として扱われていることがわかる。学習データが英語のものなのでそこまで珍しくもない漢字でも 3 token になっている。
それでも `あ` とかはそれなりに出てくるのか 1 token にまとめられている。 `日` とかは byte 単位で扱っていることがモロに出て 2 token とかになってしまう。こうなると人間的には全然わからんですな。

**Comment by yoheikikuta** [comment]
ということで、これまでの発見を全てつぎ込んだものが RoBERTa という提案モデル（と呼ぶほどのものではないかもしれないが）になっている。さらにステップ数も増やすことで、XLNet と同等以上の結果になっている。

SOTA 争いになると、本当にこれらのタスク以外でも有効なのかはわからない要素まで色々検証して頑張らなくてはいけないので大変だなぁと思う（小並）。

<p align="center">
  <img src="https://i.imgur.com/4fMm30j.png">
</p>

GLUE の各タスクの詳細や SQuAD の詳細もあるが、これは XLNet とどっこいどっこいという感じなので割愛する。RACE に関しては結構上回っている。

<p align="center">
  <img src="https://i.imgur.com/6bN6SF6.png">
</p>

一応参考情報として各タスクのリーダーボードのリンクを貼っておく。
2019/09/06 に見た範囲では確かに single model として RoBERTa がトップという感じになっている。

GLUE: https://gluebenchmark.com/leaderboard/
SQuAD: https://rajpurkar.github.io/SQuAD-explorer/
RACE:  http://www.qizhexie.com/data/RACE_leaderboard.html

**Comment by yoheikikuta** [comment]
ということで RoBERTa を読んだ。
SOTA 争いはなかなかに過酷だが、こういうのを通して本当に効いている要素とか良いモデルとかがどんどん出てくるといいですね。
